xfp/xdict.py

```diff
@@ -0,0 +1,361 @@
+        However, consistency between executions is enforced and the same Xdict will be returned each time.
+
+        ### Usage
+
+        ```python
+        from xfp import Xdict
+
+        assert Xdict({"a": 1, "b": 2}).map(lambda y, x: (f"{y}{y}", x * 10)) == Xdict({"aa": 10, "bb": 20})
+
+        collisioned = Xdict({"a": 1, "b": 2}).map(lambda _, x: ("c", x * 10))
+        assert collisioned == Xdict({"c": 20}) or collisioned == Xdict({"c": 10}) # but it will always return the same
+        ```
+        """
+        return self.from_list(self.items().map(tupled(f)))
+
+    def map_keys(self, f: Callable[[Y], E]) -> "Xdict[E, X]":
+        """Return a new Xdict, after transformation of the keys through `f`.
+
+        Transform each key with `f`, then recreate an Xdict with the result.
+        In case of conflicts in the resulted key set, which value is associated with each key is an unenforced behavior,
+        meaning this case should be avoided.
+        However, consistency between executions is enforced and the same Xdict will be returned each time.
+
+        ### Usage
+
+        ```python
+        from xfp import Xdict
+
+        assert Xdict({"a": 1, "b": 2}).map_keys(lambda y: f"{y}{y}") == Xdict({"aa": 1, "bb": 2})
+
+        collisioned = Xdict({"a": 1, "b": 2}).map(lambda _: "c")
+        assert collisioned == Xdict({"c": 2}) or collisioned == Xdict({"c": 1}) # but it will always return the same
+        ```
+        """
+        return self.map(lambda y, x: (f(y), x))
+
+    def map_values(self, f: Callable[[X], E]) -> "Xdict[Y, E]":
+        """Return a new Xdict, after transformation of the values through `f`.
+
+        Transform each value with `f`, then recreate an Xdict with the result.
+
+        ### Usage
+
+        ```python
+        from xfp import Xdict
+
+        assert Xdict({"a": 1, "b": 2}).map_values(lambda x: x * 10) == Xdict({"a": 10, "b": 20})
+        ```
+        """
+        return self.map(lambda y, x: (y, f(x)))
+
+    def filter(self, predicate: Callable[[Y, X], bool]) -> "Xdict[Y, X]":
+        """Return a new Xdict, with the couples not matching the predicate deleted.
+
+        Filter the Xdict couples (key, value) using `predicate`.
+        Couples that doesn't match are deleted.
+
+        ### Usage
+
+        ```python
+        from xfp import Xdict
+
+        assert Xdict({"a": "a", "b": "c"}).filter(lambda y, x: y == x) == Xdict({"a": "a"})
+        ```
+        """
+        return self.from_list(self.items().filter(tupled(predicate)))
+
+    def filter_keys(self, predicate: Callable[[Y], bool]) -> "Xdict[Y, X]":
+        """Return a new Xdict, with the couples not matching the predicate deleted.
+
+        Filter the Xdict keys using `predicate`.
+        Keys that doesn't match are deleted.
+
+        ### Usage
+
+        ```python
+        from xfp import Xdict
+
+        assert Xdict({"a": 1, "b": 20}).filter(lambda y: y in ["a", "c"]) == Xdict({"a": 1})
+        ```
```

Comment:
```suggestion
        ```python
            from xfp import Xdict

            assert Xdict({"a": 1, "b": 20}).filter(lambda y: y in ["a", "c"]) == Xdict({"a": 1})
        ```
```

Reply:
done